modules/core/core-repo/src/main/java/com/enonic/xp/repo/impl/vacuum/versiontable/VersionTableCleanupTask.java

```diff
@@ -181,28 +147,56 @@ private boolean versionUsedInAnyBranch( final Repository repository, final NodeV
         return false;
     }
 
-    private NodeVersionQuery createQuery( final NodeVersionId lastVersionId )
+    private NodeVersionQuery createQuery( final Repository repository )
     {
         final NodeVersionQuery.Builder builder = NodeVersionQuery.create();
 
-        if ( lastVersionId != null )
+        builder.addQueryFilter( RangeFilter.create().
+            fieldName( VersionIndexPath.TIMESTAMP.getPath() ).
+            to( ValueFactory.newDateTime( this.until ) ).
+            build() );
+
+        builder.size( -1 ).
+            batchSize( 100 ).
+            batchCallback( result -> this.handleBatchCallback( (SearchResult) result, repository, builder.build() ) );
+
+        return builder.build();
+    }
+
+    private void handleBatchCallback( final SearchResult searchResult, final Repository repository,
+                                      final NodeVersionQuery nodeVersionQuery )
+    {
+        final List<NodeVersionDocumentId> toBeDeleted = Lists.newArrayList();
+
+        final long versionTotal = searchResult.getTotalHits();
+        if ( !this.repositoryCleanStarted )
         {
-            final RangeFilter versionIdFilter = RangeFilter.create().
-                fieldName( VersionIndexPath.VERSION_ID.getPath() ).
-                from( ValueFactory.newString( lastVersionId.toString() ) ).
-                build();
-            builder.addQueryFilter( versionIdFilter );
+            if ( listener != null )
+            {
+                listener.vacuumingVersionRepository( repository.getId(), versionTotal );
+
+            }
+            this.repositoryCleanStarted = true;
         }
 
-        final RangeFilter mustBeOlderThanFilter = RangeFilter.create().
-            fieldName( VersionIndexPath.TIMESTAMP.getPath() ).
-            to( ValueFactory.newDateTime( this.until ) ).
-            build();
+        if ( searchResult.getHits().getSize() == 0 )
```

Comment:
will be better to use `isEmpty`

Reply:
There is no such method in `SearchHits` class. Not sure that is need to be added just for one call.